File: rust/crates/corlinman-gateway/src/middleware/tenant_scope.rs
//! Tenant-scoping middleware for `/admin/*`.
//!
//! Phase 4 W1 4-1A Item 3. Resolves an inbound admin request to a single
//! [`TenantId`] and stashes it in the request extensions so handlers can
//! pull it back via the [`Tenant`] extractor. Two policy modes, decided
//! once at boot from the `[tenants].enabled` config switch:
//!
//! - **Disabled (legacy single-tenant)**: every request gets
//!   [`TenantId::legacy_default`] without consulting the request — the
//!   middleware is effectively transparent. This is the byte-for-byte
//!   pre-Phase-4 behaviour.
//! - **Enabled (multi-tenant)**: the middleware extracts a candidate
//!   slug from `?tenant=<slug>` (Phase 4 W1 only — session-cookie tenant
//!   claims land in a Phase 4 W1 follow-up alongside the
//!   `AdminSessionStore` schema bump). The slug is parsed through
//!   [`TenantId::new`] and validated against the operator-allowed set
//!   stored on `TenantScopeState`. Empty / missing query falls back to
//!   the configured default tenant.
//!
//! Two short-circuit error paths:
//!
//! - HTTP **400 `invalid_tenant_slug`** when the query carries a slug
//!   that fails the `^[a-z][a-z0-9-]{0,62}$` shape — surfaced clearly so
//!   the UI can distinguish a typo from an authorisation failure.
//! - HTTP **403 `tenant_not_allowed`** when the slug parses but is not
//!   in `state.allowed`. This is the security boundary that satisfies
//!   the Wave 1 acceptance: an operator scoped to tenant A who tries
//!   `?tenant=B` sees a 403, not a silently empty result.
//!
//! Mount order: this layer sits *inside* `require_admin` (so anonymous
//! callers never see `tenant_not_allowed`; they get 401 first) and
//! *outside* the per-route handlers (so handlers always observe a
//! resolved `TenantId` in extensions).

use std::collections::BTreeSet;
use std::sync::Arc;

use axum::{
    body::Body,
    extract::{FromRequestParts, State},
    http::{request::Parts, Request, StatusCode},
    middleware::Next,
    response::{IntoResponse, Response},
    Json,
};
use corlinman_tenant::TenantId;
use serde_json::json;

/// Boot-time state for [`tenant_scope`]. `enabled` mirrors
/// `Config::tenants::enabled` at gateway start; `allowed` is the union
/// of `[tenants].allowed` slugs (validated at boot) plus
/// [`TenantId::legacy_default`]; `fallback` is the tenant returned
/// when the request omits `?tenant=` (matches `[tenants].default`).
///
/// Cloneable: every field is small or `Arc`-wrapped so the layer
/// can be stamped onto the admin router without per-clone allocation
/// pressure. Held by value on the layer; not stored in `AdminState`.
#[derive(Clone)]
pub struct TenantScopeState {
    pub enabled: bool,
    pub allowed: Arc<BTreeSet<TenantId>>,
    pub fallback: TenantId,
}

impl TenantScopeState {
    /// Convenience: build a disabled state where every request resolves
    /// to [`TenantId::legacy_default`]. Used by tests that want to
    /// assert handler behaviour without exercising tenant scoping.
    pub fn disabled() -> Self {
        Self {
            enabled: false,
            allowed: Arc::new(BTreeSet::new()),
            fallback: TenantId::legacy_default(),
        }
    }
}

/// Axum extractor: pulls the tenant id resolved by [`tenant_scope`]
/// out of request extensions. Returns 500 if missing — that's a wiring
/// bug (handler used before the middleware mounted), surfaced loudly
/// so it doesn't degrade silently to "default".
#[derive(Clone, Debug)]
pub struct Tenant(pub TenantId);

#[axum::async_trait]
impl<S> FromRequestParts<S> for Tenant
where
    S: Send + Sync,
{
    type Rejection = (StatusCode, Json<serde_json::Value>);

    async fn from_request_parts(parts: &mut Parts, _state: &S) -> Result<Self, Self::Rejection> {
        match parts.extensions.get::<TenantId>().cloned() {
            Some(t) => Ok(Tenant(t)),
            None => Err((
                StatusCode::INTERNAL_SERVER_ERROR,
                Json(json!({
                    "error": "tenant_extension_missing",
                    "hint": "tenant_scope middleware was not mounted before this handler",
                })),
            )),
        }
    }
}

/// Resolve the tenant for this request and stash it in extensions.
///
/// See module docs for the policy. On success the inner handler
/// observes `req.extensions().get::<TenantId>()` populated; on policy
/// failure the middleware short-circuits with 400 / 403 and the
/// handler never runs.
pub async fn tenant_scope(
    State(state): State<TenantScopeState>,
    mut req: Request<Body>,
    next: Next,
) -> Response {
    let tenant = if !state.enabled {
        state.fallback.clone()
    } else {
        match extract_tenant_query(req.uri().query().unwrap_or("")) {
            None => state.fallback.clone(),
            Some(raw) => match TenantId::new(&raw) {
                Err(err) => {
                    return (
                        StatusCode::BAD_REQUEST,
                        Json(json!({
                            "error": "invalid_tenant_slug",
                            "slug": raw,
                            "reason": err.to_string(),
                        })),
                    )
                        .into_response();
                }
                Ok(t) => {
                    if !state.allowed.contains(&t) {
                        return (
                            StatusCode::FORBIDDEN,
                            Json(json!({
                                "error": "tenant_not_allowed",
                                "slug": t.as_str(),
                            })),
                        )
                            .into_response();
                    }
                    t
                }
            },
        }
    };

    req.extensions_mut().insert(tenant);
    next.run(req).await
}

/// Extract the first `tenant=` value from a percent-encoded query
/// string. Returns the raw value (URL-decoded) without further
/// validation — the caller is responsible for parsing through
/// [`TenantId::new`].
///
/// Inlined rather than pulled from a query-parser crate to keep this
/// middleware free of new dependencies. The tenant query is single-
/// valued, the slug regex bans `=` and `&`, and we only need the
/// first match — so a hand-rolled scan over `&`-split pairs is enough.
fn extract_tenant_query(q: &str) -> Option<String> {
    for pair in q.split('&') {
        let mut it = pair.splitn(2, '=');
        let key = it.next()?;
        if key != "tenant" {
            continue;
        }
        let raw = it.next().unwrap_or("");
        // Minimal percent-decode for the characters slugs can contain.
        // Slugs are `[a-z0-9-]`, none of which need percent-encoding,
        // so a naive URL-decode is overkill; we just unescape `%2D`
        // (`-`) defensively in case a client over-encodes it. Anything
        // exotic falls through and `TenantId::new` will reject it.
        let unescaped = raw.replace("%2D", "-").replace("%2d", "-");
        return Some(unescaped);
    }
    None
}

#[cfg(test)]
mod tests {
    use super::*;
    use axum::body::to_bytes;
    use axum::routing::get;
    use axum::{middleware, Router};
    use tower::ServiceExt;

    /// Handler that pulls the tenant out via the extractor and echoes
    /// it back as JSON. Lets the tests assert on the resolved id end-
    /// to-end instead of poking extensions directly.
    async fn echo(Tenant(t): Tenant) -> Json<serde_json::Value> {
        Json(json!({"tenant": t.as_str()}))
    }

    fn app(state: TenantScopeState) -> Router {
        Router::new()
            .route("/probe", get(echo))
            .layer(middleware::from_fn_with_state(state, tenant_scope))
    }

    fn allowed(slugs: &[&str]) -> Arc<BTreeSet<TenantId>> {
        let mut s = BTreeSet::new();
        s.insert(TenantId::legacy_default());
        for slug in slugs {
            s.insert(TenantId::new(*slug).expect("test slug must be valid"));
        }
        Arc::new(s)
    }

    #[tokio::test]
    async fn disabled_resolves_every_request_to_default() {
        let app = app(TenantScopeState::disabled());
        let resp = app
            .oneshot(
                Request::builder()
                    .uri("/probe?tenant=acme")
                    .body(Body::empty())
                    .unwrap(),
            )
            .await
            .unwrap();
        assert_eq!(resp.status(), StatusCode::OK);
        let bytes = to_bytes(resp.into_body(), 1024).await.unwrap();
        let body: serde_json::Value = serde_json::from_slice(&bytes).unwrap();
        assert_eq!(body["tenant"], "default");
    }

    #[tokio::test]
    async fn enabled_falls_back_to_configured_default_when_query_absent() {
        let state = TenantScopeState {
            enabled: true,
            allowed: allowed(&["acme"]),
            fallback: TenantId::legacy_default(),
        };
        let app = app(state);
        let resp = app
            .oneshot(Request::builder().uri("/probe").body(Body::empty()).unwrap())
            .await
            .unwrap();
        assert_eq!(resp.status(), StatusCode::OK);
        let body: serde_json::Value =
            serde_json::from_slice(&to_bytes(resp.into_body(), 1024).await.unwrap()).unwrap();
        assert_eq!(body["tenant"], "default");
    }

    #[tokio::test]
    async fn enabled_resolves_allowed_tenant_query() {
        let state = TenantScopeState {
            enabled: true,
            allowed: allowed(&["acme", "bravo"]),
            fallback: TenantId::legacy_default(),
        };
        let app = app(state);
        let resp = app
            .oneshot(
                Request::builder()
                    .uri("/probe?tenant=bravo")
                    .body(Body::empty())
                    .unwrap(),
            )
            .await
            .unwrap();
        assert_eq!(resp.status(), StatusCode::OK);
        let body: serde_json::Value =
            serde_json::from_slice(&to_bytes(resp.into_body(), 1024).await.unwrap()).unwrap();
        assert_eq!(body["tenant"], "bravo");
    }

    #[tokio::test]
    async fn enabled_rejects_invalid_slug_with_400() {
        let state = TenantScopeState {
            enabled: true,
            allowed: allowed(&["acme"]),
            fallback: TenantId::legacy_default(),
        };
        let app = app(state);
        let resp = app
            .oneshot(
                Request::builder()
                    .uri("/probe?tenant=BAD!!")
                    .body(Body::empty())
                    .unwrap(),
            )
            .await
            .unwrap();
        assert_eq!(resp.status(), StatusCode::BAD_REQUEST);
        let body: serde_json::Value =
            serde_json::from_slice(&to_bytes(resp.into_body(), 1024).await.unwrap()).unwrap();
        assert_eq!(body["error"], "invalid_tenant_slug");
    }

    /// The Wave 1 acceptance line: an operator targeting a tenant they
    /// have not been granted access to gets a 403, not a silently
    /// empty list.
    #[tokio::test]
    async fn enabled_rejects_disallowed_tenant_with_403() {
        let state = TenantScopeState {
            enabled: true,
            allowed: allowed(&["acme"]),
            fallback: TenantId::legacy_default(),
        };
        let app = app(state);
        let resp = app
            .oneshot(
                Request::builder()
                    .uri("/probe?tenant=bravo")
                    .body(Body::empty())
                    .unwrap(),
            )
            .await
            .unwrap();
        assert_eq!(resp.status(), StatusCode::FORBIDDEN);
        let body: serde_json::Value =
            serde_json::from_slice(&to_bytes(resp.into_body(), 1024).await.unwrap()).unwrap();
        assert_eq!(body["error"], "tenant_not_allowed");
        assert_eq!(body["slug"], "bravo");
    }

    #[tokio::test]
    async fn missing_extension_returns_500_explicit_wiring_bug() {
        // Direct extractor call without the middleware — proves the
        // 500 path fires when the layer is not mounted, so a future
        // refactor that drops the layer fails loudly.
        async fn handler(t: Result<Tenant, (StatusCode, Json<serde_json::Value>)>) -> StatusCode {
            match t {
                Ok(_) => StatusCode::OK,
                Err((code, _)) => code,
            }
        }
        let app = Router::new().route("/probe", get(handler));
        let resp = app
            .oneshot(Request::builder().uri("/probe").body(Body::empty()).unwrap())
            .await
            .unwrap();
        assert_eq!(resp.status(), StatusCode::INTERNAL_SERVER_ERROR);
    }

    #[test]
    fn extract_tenant_query_finds_first_match() {
        assert_eq!(extract_tenant_query("tenant=acme"), Some("acme".into()));
        assert_eq!(
            extract_tenant_query("foo=1&tenant=bravo&bar=2"),
            Some("bravo".into())
        );
        assert_eq!(extract_tenant_query("foo=1&bar=2"), None);
        assert_eq!(extract_tenant_query(""), None);
        // Tolerates over-encoded `-` even though slugs don't strictly
        // need it.
        assert_eq!(
            extract_tenant_query("tenant=ac%2Dme"),
            Some("ac-me".into())
        );
    }
}

File: python/packages/corlinman-server/src/corlinman_server/gateway/middleware/tenant_scope.py
"""Tenant-scoping middleware for ``/admin/*`` (and any other gated surface).

Python port of ``rust/crates/corlinman-gateway/src/middleware/tenant_scope.rs``.

Two policy modes, decided once at boot from the ``[tenants].enabled``
config switch:

* **Disabled (legacy single-tenant)** — every request gets
  :func:`~corlinman_server.tenancy.default_tenant` without consulting
  the request. The middleware is effectively transparent. This is the
  byte-for-byte pre-Phase-4 behaviour.
* **Enabled (multi-tenant)** — the middleware extracts a candidate
  slug from ``?tenant=<slug>`` (or the ``X-Corlinman-Tenant`` header
  fallback for clients that can't put query params in the URL). The
  slug is parsed through :meth:`TenantId.new` and validated against
  the operator-allowed set in :attr:`TenantScopeState.allowed`. Empty
  / missing falls back to :attr:`TenantScopeState.fallback`.

Two short-circuit error paths mirror the Rust impl exactly:

* HTTP **400 ``invalid_tenant_slug``** when the query carries a slug
  that fails the ``^[a-z][a-z0-9-]{0,62}$`` shape.
* HTTP **403 ``tenant_not_allowed``** when the slug parses but is not
  in the allowed set.

Mount order: this layer sits *inside* :class:`AdminAuthMiddleware` (so
anonymous callers never see ``tenant_not_allowed`` — they get 401 first)
and *outside* per-route handlers (so handlers always observe a resolved
:class:`TenantId` on ``request.state.tenant``).
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any
from urllib.parse import parse_qsl, unquote

import structlog
from fastapi import Depends, HTTPException, Request, status
from starlette.middleware.base import BaseHTTPMiddleware
from starlette.responses import JSONResponse, Response
from starlette.types import ASGIApp

from corlinman_server.tenancy import (
    TenantId,
    TenantIdError,
    default_tenant,
)

logger = structlog.get_logger(__name__)


#: Header clients may use as an alternative to ``?tenant=<slug>`` when a
#: query parameter is awkward (e.g. websocket upgrade requests). Read
#: only when the query is absent so query-based scoping stays
#: authoritative.
TENANT_HEADER_NAME: str = "X-Corlinman-Tenant"

#: Path key middleware can also pull a tenant slug from, e.g. routes
#: mounted under ``/v1/tenants/{tenant}/...``. Activates when the
#: matched route exposes the parameter in ``request.path_params``.
TENANT_PATH_PARAM: str = "tenant"


# ---------------------------------------------------------------------------
# State
# ---------------------------------------------------------------------------


@dataclass
class TenantScopeState:
    """Boot-time state for the tenant-scope middleware.

    ``enabled`` mirrors ``Config::tenants::enabled`` at gateway start;
    ``allowed`` is the union of ``[tenants].allowed`` slugs plus the
    legacy default; ``fallback`` is the tenant returned when the request
    omits a tenant indicator (matches ``[tenants].default``).
    """

    enabled: bool = False
    allowed: frozenset[TenantId] = field(default_factory=frozenset)
    fallback: TenantId = default_tenant()

    @classmethod
    def disabled(cls) -> "TenantScopeState":
        """Build a disabled state where every request resolves to
        :func:`~corlinman_server.tenancy.default_tenant`. Used by tests
        that want to assert handler behaviour without exercising tenant
        scoping."""

        return cls(enabled=False, allowed=frozenset(), fallback=default_tenant())


# ---------------------------------------------------------------------------
# Helpers — pure, exported for tests.
# ---------------------------------------------------------------------------


def extract_tenant_query(query: str) -> str | None:
    """Pull the first ``tenant=`` value out of a percent-encoded query
    string.

    Mirrors the Rust ``extract_tenant_query`` helper: the tenant query
    is single-valued, slugs are ``[a-z0-9-]`` (none of which need
    percent-encoding), so we just defensively unescape ``%2D`` /
    ``%2d``. Anything more exotic falls through and
    :meth:`TenantId.new` rejects it.
    """

    if not query:
        return None
    # ``parse_qsl`` keeps order and respects ``&`` separators.
    for key, value in parse_qsl(query, keep_blank_values=True):
        if key == "tenant":
            return unquote(value).replace("%2D", "-").replace("%2d", "-")
    return None


def _candidate_slug(request: Request) -> str | None:
    """Find a tenant slug on the request, in priority order:

    1. ``?tenant=<slug>`` query string.
    2. ``X-Corlinman-Tenant`` header.
    3. Matched ``{tenant}`` path parameter.

    Returns ``None`` when none of the three sources carries a value.
    """

    raw_query = request.url.query
    q = extract_tenant_query(raw_query) if raw_query else None
    if q:
        return q

    hdr = request.headers.get(TENANT_HEADER_NAME)
    if hdr:
        return hdr.strip()

    path_value = request.path_params.get(TENANT_PATH_PARAM)
    if isinstance(path_value, str) and path_value:
        return path_value

    return None


def _resolve_state(request: Request) -> TenantScopeState | None:
    state = getattr(request.app.state, "tenant_scope", None)
    if isinstance(state, TenantScopeState):
        return state
    return None


# ---------------------------------------------------------------------------
# Middleware
# ---------------------------------------------------------------------------


class TenantScopeMiddleware(BaseHTTPMiddleware):
    """Resolve the inbound request's tenant and stash it on
    ``request.state.tenant``.

    See module docs for the policy. On success the inner handler sees
    ``request.state.tenant`` populated; on policy failure the
    middleware short-circuits with 400 / 403 and the handler never runs.
    """

    def __init__(
        self,
        app: ASGIApp,
        state: TenantScopeState | None = None,
    ) -> None:
        super().__init__(app)
        self._state = state or TenantScopeState.disabled()

    async def dispatch(
        self,
        request: Request,
        call_next: Any,
    ) -> Response:
        state = _resolve_state(request) or self._state

        # Skip resolution when something upstream (e.g. ApiKeyAuthMiddleware)
        # already pinned a tenant; that upstream value is the source of
        # truth — re-resolving here could surface a 403 for a path the
        # caller never opted into.
        existing = getattr(request.state, "tenant", None)
        if isinstance(existing, TenantId):
            return await call_next(request)

        if not state.enabled:
            request.state.tenant = state.fallback
            return await call_next(request)

        raw = _candidate_slug(request)
        if raw is None or raw == "":
            request.state.tenant = state.fallback
            return await call_next(request)

        try:
            tenant = TenantId.new(raw)
        except TenantIdError as exc:
            return JSONResponse(
                {
                    "error": "invalid_tenant_slug",
                    "slug": raw,
                    "reason": str(exc),
                },
                status_code=status.HTTP_400_BAD_REQUEST,
            )

        if tenant not in state.allowed:
            return JSONResponse(
                {
                    "error": "tenant_not_allowed",
                    "slug": tenant.as_str(),
                },
                status_code=status.HTTP_403_FORBIDDEN,
            )

        request.state.tenant = tenant
        return await call_next(request)


def install_tenant_scope_middleware(
    app: Any,
    *,
    enabled: bool = False,
    allowed: frozenset[TenantId] | set[TenantId] | None = None,
    fallback: TenantId | None = None,
) -> TenantScopeState:
    """Attach :class:`TenantScopeMiddleware` to ``app``.

    Returns the :class:`TenantScopeState` so the caller can swap
    ``allowed`` / ``fallback`` later (config reload). The same instance
    is published on ``app.state.tenant_scope``.
    """

    state = TenantScopeState(
        enabled=enabled,
        allowed=frozenset(allowed or ()),
        fallback=fallback or default_tenant(),
    )
    app.state.tenant_scope = state
    app.add_middleware(TenantScopeMiddleware, state=state)
    return state


# ---------------------------------------------------------------------------
# FastAPI ``Depends`` factory
# ---------------------------------------------------------------------------


def require_tenant() -> Any:
    """Per-route extractor returning the resolved :class:`TenantId`.

    Usage::

        @router.get("/admin/things")
        async def handler(tenant: TenantId = Depends(require_tenant())):
            ...

    Returns whatever the middleware stashed on
    ``request.state.tenant``. Raises HTTP 500 with a clear wiring-bug
    hint when the middleware wasn't mounted (matches the Rust
    extractor's ``tenant_extension_missing`` envelope).
    """

    def dependency(request: Request) -> TenantId:
        tenant = getattr(request.state, "tenant", None)
        if isinstance(tenant, TenantId):
            return tenant
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail={
                "error": "tenant_extension_missing",
                "hint": "tenant_scope middleware was not mounted before this handler",
            },
        )

    return Depends(dependency)


__all__ = [
    "TENANT_HEADER_NAME",
    "TENANT_PATH_PARAM",
    "TenantScopeMiddleware",
    "TenantScopeState",
    "extract_tenant_query",
    "install_tenant_scope_middleware",
    "require_tenant",
]
